### ISSUE TITLE
Support forward slash in Windows paths

### ISSUE THREAD

**Issue by EmilGedda** [nvim-neo-tree/neo-tree.nvim#933]
Windows accepts forward slashes as path component separators in most of their API's (and in explorer.exe as well as cmd.exe). It would be nice if neo-tree.nvim did the same. 

This does not work on Windows: 
```lua
require("neo-tree.command").execute({ toggle = true, dir = "foo/bar" })
```
Looks like it tries to open a non-existant path.

This does work:
```lua
require("neo-tree.command").execute({ toggle = true, dir = "foo\\bar" })
```
And the contents of `bar` is shown.

**Comment by pysan3** [comment]
Hey @miversen33 

Is there any change this issue can be solved with (or something similar to) #1023 ?

**Comment by miversen33** [comment]
So the gist of this request is basically, whatever function actually resolves the path, it should accept this kind of pathing as well (though do we make it only accept this path if the os is windows? Idk).

As an alternative, we could cheese it and have the `filesystem.navigate` function just convert all `foo/bar` references to `foo\\bar` references (basically convert `/` -> `\\` behind the scenes). I don't imagine it would be tough to do either. Just adding the following line

```lua
state.path = state.path:gsub('\', '//')
```

to [lua/neo-tree/sources/filesystem.init#L137](https://github.com/nvim-neo-tree/neo-tree.nvim/blob/7ce4958108c52f764c32605c61cb6d0c5943161c/lua/neo-tree/sources/filesystem/init.lua#L137)

Thinking about it, I don't actually hate that idea. @cseickel thoughts on this?

**Comment by cseickel** [comment]
> So the gist of this request is basically, whatever function actually resolves the path, it should accept this kind of pathing as well (though do we make it only accept this path if the os is windows? Idk).
> 
> As an alternative, we could cheese it and have the `filesystem.navigate` function just convert all `foo/bar` references to `foo\\bar` references (basically convert `/` -> `\\` behind the scenes). I don't imagine it would be tough to do either. Just adding the following
> 
> ```lua
> state.path = state.path:gsub('\', '//')
> path_to_reveal = path_to_reveal:gsub('\', '//')
> path = path:gsub('\', '//')
> ```
> 
> to [lua/neo-tree/sources/filesystem.init#L120](https://github.com/nvim-neo-tree/neo-tree.nvim/blob/7ce4958108c52f764c32605c61cb6d0c5943161c/lua/neo-tree/sources/filesystem/init.lua#L120)
> 
> Thinking about it, I don't actually hate that idea. @cseickel thoughts on this?

I think you have your slashes reversed. The request is to support **forward** slashes, unix style. Your suggestion is about escaping **back**slashes.

I'm guessing there is a branch of code somewhere that interprets a path based on windows conventions if the os is windows which is not recognizing the use of a unix style path separator. I'd need more details on where the problem is before I can make any suggestions.

I honestly never knew that windows would accept both unix style and windows style separators. It seems completely insane to me. What happens if you use both? Does backslash become an escape character or an alternate path separator?

**Comment by miversen33** [comment]
> I think you have your slashes reversed. The request is to support forward slashes, unix style. Your suggestion is about escaping backslashes.

Not quite. I am replacing forward slashes with escaped backslashes. This means a user can request directory `foo/bar` and the filesystem source will see it as `foo\\bar` (which is what it expects currently I believe). It seems _a bit_ hacky, but unless I am mistaken, it shouldn't cause any issues with `vim.loop` (btw this is now [vim.uv](https://neovim.io/doc/user/lua.html#vim.uv)) or plenary. It is essentially forcing consistent paths before we start messing about with the path string

> I honestly never knew that windows would accept both unix style and windows style separators. It seems completely insane to me. What happens if you use both? Does backslash become an escape character or an alternate path separator?

So fun fact, apparently windows treats `/` and `\` the same. I can (for example) navigate to `C:/Users\miver` and it works just the same as `C:\Users\miver` and `C:/Users/miver`. Tested quick in powershell, command prompt and the file explorer. It seems NTFS doesn't care.

Because who needs consistency right? :upside_down_face:

**Comment by pysan3** [comment]
If I understand correctly, in `cmd` the escape char is `^` and in powershell, it's `<backtick>` (damn how do I put a backtick in markdown lol)

Both cases, `\` does not mean to escape something and is used for path separator.

Neovim's core team seems to prefer using `/` in Windows: https://github.com/neovim/neovim/issues/15220 (link is broken in that issue, see [this commit](https://github.com/neovim/neovim/commit/1c3ca4f18fdc403813d8959b49626ac1c99e2c59) for details).

But old vim functions prefer backslashes. `:=vim.fn.getcwd() => C:\...` so now it's a complete mess.

**Comment by cseickel** [comment]
> Because who needs consistency right? 🙃

> so now it's a complete mess.

100%

@miversen33 I think you are on the right track, but the place to enforce that consistency is probably where the user input is first taken in, in the command module. Maybe in this function?
https://github.com/nvim-neo-tree/neo-tree.nvim/blob/7f6fa04dbd8e8c79d1af33bc90e856b65d8641da/lua/neo-tree/command/parser.lua#L81

**Comment by pysan3** [comment]
I noticed this inconsistency when digging into #1026 and I'm still not sure why there hasn't been much problem in neo-tree's code yet.

### 1. Opening file.

Let's try to open `C:/Users/User/AppData/Local/nvim/lua/target{1..5}.lua` while `CWD = C:/Users/User/AppData/Local/nvim`.
There's a lot of ways to do it.

1. `:e lua/target1.lua` or `:e lua\target1.lua`
2. `:e C:/Users/User/AppData/Local/nvim/lua/target2.lua` (Neotree uses this I think?)
3. `:e C:\Users\User\AppData\Local\nvim\lua\target3.lua`
2. `:e ~/AppData/Local/nvim/lua/target4.lua`
3. `:e ~\AppData\Local\nvim\lua\target5.lua`

### 2. Look at `:ls`

The results of files opened with each command correspondes to below.

1. `lua\target1.lua`
2. `~/AppData/Local/nvim/lua/target2.lua`
3. `~\AppData\Local\nvim\lua\target3.lua`
2. `~/AppData/Local/nvim/lua/target4.lua`
3. `~\AppData\Local\nvim\lua\target5.lua`

So, as you can see, `C:/Users/User` is always substituted by `~` but the type of slashes does matter.

### 3. Look at `vim.fn.expand("%:p")`

1. `C:\Users\User\AppData\Local\nvim\lua\target1.lua`
2. `C:\Users\User/AppData/Local/nvim/lua/target2.lua` !!!
3. `C:\Users\User\AppData\Local\nvim\lua\target3.lua`
2. `C:\Users\User/AppData/Local/nvim/lua/target4.lua` !!!
3. `C:\Users\User\AppData\Local\nvim\lua\target5.lua`

Now, here's where it starts to get weird.
The result of 2 and 4 have `/` and `\\` mixed up because of the `:p`.

### 4. `mksession!`

Let's save the current session with `:mks!` which results in a file named `Session.vim` right inside `$CWD`.

When you peek inside this file, you'll see lines that prints as below, which has all backslashes replaced with `/`,
and uses `~`.

```txt
badd +1 `lua\target1.lua`
badd +1 `~/AppData/Local/nvim/lua/target2.lua`
badd +1 `~/AppData/Local/nvim/lua/target3.lua`
badd +1 `~/AppData/Local/nvim/lua/target4.lua`
badd +1 `~/AppData/Local/nvim/lua/target5.lua`
```

This behavior is however different from `vim` which does not substitute backslashes by default.
Introduced in [this commit](https://github.com/neovim/neovim/commit/1c3ca4f18fdc403813d8959b49626ac1c99e2c59),
[related issue](https://github.com/neovim/neovim/issues/15220).
(Use `:h sessionoptions` to change the behavior. `nvim` deletes this option and make `/` the default.)

### `vim.fn.getcwd()`

This however returns path with all backslashes.

The biggest issue that should be caused by this inconsistency is [`is_subpath`](https://github.com/nvim-neo-tree/neo-tree.nvim/blob/7f6fa04dbd8e8c79d1af33bc90e856b65d8641da/lua/neo-tree/utils.lua#L674) where `string.sub(path, 1, string.len(base)) == base` might not be `true` due to difference in slashes.

However neo-tree is surprisingly working as expected in most cases probably thanks to [`windowize_path`](https://github.com/nvim-neo-tree/neo-tree.nvim/blame/7f6fa04dbd8e8c79d1af33bc90e856b65d8641da/lua/neo-tree/utils.lua#L896C7-L896C7) but there are certain places where this is not used but should be critical. Like [here](https://github.com/nvim-neo-tree/neo-tree.nvim/blob/7f6fa04dbd8e8c79d1af33bc90e856b65d8641da/lua/neo-tree/sources/manager.lua#L150). I'm quite confused now lol.

@cseickel Maybe we should wrap all calls to `vim.fn.expand, getcwd, etc` and `vim.api.nvim_buf_get_name` and substitute backslashes to `/` and force to use forwardslash throughout the project? Or shall we simply keep it as-is?

**Comment by miversen33** [comment]
> I think you are on the right track, but the place to enforce that consistency is probably where the user input is first taken in, in the command module. Maybe in this function?

> Maybe we should wrap all calls to vim.fn.expand, getcwd, etc and vim.api.nvim_buf_get_name and substitute backslashes to / and force to use forwardslash throughout the project? Or shall we simply keep it as-is?

By doing this, we (as Neo-tree) are making the active decision that we will sanitize paths before passing them to the underlying source to handle. Not to say that's a bad thing, just a decision that I didn't feel comfortable making (which is why I moved the suggestion into the filesystem source instead of above in the command resolution, which is where I was initially looking).

IMO, while taking this approach may be the best decision, it might be wise to take it as a 2 step approach. Resolving this direct issue (allowing forward slashes in the path), and then targeting path sanitation/fixing for `3.x`

My biggest concern is that one of these approaches is _much_ smaller and easier to validate/verify than the other :)

**Comment by cseickel** [comment]
> By doing this, we (as Neo-tree) are making the active decision that we will sanitize paths before passing them to the underlying source to handle.

That's a good point. I didn't consider that someone could be on windows but still be passing an argument to a source that interacts with a *nix host such as a docker container or remote host.

Given that, your suggestion of handling within the filesystem source is a good fix for this specific issue.

Longer term we should handle paths in a consistent manner throughout the plugin and external sources. As @pysan3 pointed out, the issue is complicated and there are likely many edge cases waiting to be found within our codebase. I think we should drop all representations of paths as strings and replace them with plenary `Path` objects or come up with our own abstraction that can handle the windows/nix differences properly. Hopefully Plenary's implementation works or we can get fixes in where needed. This might be a v4.x change.

**Comment by miversen33** [comment]
> That's a good point. I didn't consider that someone could be on windows but still be passing an argument to a source that interacts with a *nix host such as a docker container or remote host.

I appreciate you :)

I'll slap together a PR for this this evening, the change is really small so it shouldn't take me too long to do and test.

> I think we should drop all representations of paths as strings and replace them with plenary Path objects or come up with our own abstraction that can handle the windows/nix differences properly. Hopefully Plenary's implementation works or we can get fixes in where needed

IIRC Plenary Path's are a touch buggy but I haven't touched them since I was playing around with Telescope back in the day. I do think letting pathing be someone else's problem is a great solution. Properly handling paths is hard.

**Comment by WillEhrendreich** [comment]
I was trying to look into this a little bit, because it seemed insane and messy. surely enough, that's right, it is insane and messy. I'm totally for the sanitizing everything with the vim.fs.normalize approach, though IIRC there's some interesting problems and discussion around that subject in neovim recently.. 

ok here's the pr.. 

it looks like using this to sanitize might be ok now? https://github.com/neovim/neovim/commit/8a7e3353eb5bffb10015254917361266b4b20511

**Comment by miversen33** [comment]
Quick test with `vim.fs.normalize` nets the below results
```vim
:lua print(vim.fs.normalize("C:\\Users/miver"))
C:/Users/miver
```

I think we might be able to get away with using normalize in yes. It will take more testing to be sure it works as expected generally though :)

**Comment by pysan3** [comment]
Could you also stress test non ascii chars as well as emojis included in the path? 

We might also have to check how the git related codes work since Windows Git uses a weird encoding for non-ascii characters. 

This just came to my mind but would it be fixed if we execute `chcp 65001` every time before the git command? I don't have a Windows PC right now so I'm sorry to bother you.

**Comment by miversen33** [comment]
> Could you also stress test non ascii chars as well as emojis included in the path?
> 
> We might also have to check how the git related codes work since Windows Git uses a weird encoding for non-ascii characters.

I think we are probably safe. I made a truly horrendous discovery in that NTFS allows emojis in your path
![image](https://github.com/nvim-neo-tree/neo-tree.nvim/assets/2640668/12570371-eda3-4801-9484-ca6c3e629485)
![image](https://github.com/nvim-neo-tree/neo-tree.nvim/assets/2640668/1bd37da8-2437-4586-913e-decd33fe477e)

Using `vim.fn.normalize` seems to be the best way to handle this whole path escaping thing IMO

> This just came to my mind but would it be fixed if we execute chcp 65001 every time before the git command? I don't have a Windows PC right now so I'm sorry to bother you.

I have no idea what this is so I can't really speak to it unfortunately

**Comment by WillEhrendreich** [comment]
> On my linux machine (where the path C:\ is nonsensical) and the function still cleaned it up to something that is "standard". So it is most likely (speculating as I don't feel like digging into the vim source to verify) just running some string regex replaces to make the path "normal" (hence the function name).

I will fetch dis. 

copy-la-pasta, here's the code fasta!  

from https://github.com/neovim/neovim/blob/a3f4598226c4d01e4fbc41181a1ad21793862fe3/runtime/lua/vim/fs.lua#L331
```lua

--- Normalize a path to a standard format. A tilde (~) character at the
--- beginning of the path is expanded to the user's home directory and any
--- backslash (\\) characters are converted to forward slashes (/). Environment
--- variables are also expanded.
---
--- Examples:
--- <pre>lua
---   vim.fs.normalize('C:\\\\Users\\\\jdoe')
---   --> 'C:/Users/jdoe'
---
---   vim.fs.normalize('~/src/neovim')
---   --> '/home/jdoe/src/neovim'
---
---   vim.fs.normalize('$XDG_CONFIG_HOME/nvim/init.vim')
---   --> '/Users/jdoe/.config/nvim/init.vim'
--- </pre>
---
---@param path (string) Path to normalize
---@param opts table|nil Options:
---             - expand_env: boolean Expand environment variables (default: true)
---@return (string) Normalized path
function M.normalize(path, opts)
  opts = opts or {}

  vim.validate({
    path = { path, { 'string' } },
    expand_env = { opts.expand_env, { 'boolean' }, true },
  })

  if path:sub(1, 1) == '~' then
    local home = vim.uv.os_homedir() or '~'
    if home:sub(-1) == '\\' or home:sub(-1) == '/' then
      home = home:sub(1, -2)
    end
    path = home .. path:sub(2)
  end

  if opts.expand_env == nil or opts.expand_env then
    path = path:gsub('%$([%w_]+)', vim.uv.os_getenv)
  end

  return (path:gsub('\\', '/'):gsub('/+', '/'):gsub('(.)/$', '%1'))
end

```  

yeah, it's simply transforming it, not making sure it exists. It does do some expansion of env vars though, so that's cool. I don't know if that helps, but it's cool. 

shocking about the emoji thing.. 🤷‍♂️

I'm envisioning a dystopian future full of filesystems that look like
C:/🎁/👻/😆/🐱‍👤/🏎/🥳/(╯°□°）╯︵ ┻━┻/Q2 fiscal year 1323/straightLaced_biznassREPORTZ.🦅

we will soon have to support the .🦅 file extension.. you have been warned.

**Comment by miversen33** [comment]
I've submitted PR #1051 that should resolve this. I tested on my windows and linux machines and it appears to be operating "as expected".

@cseickel fun side note, Netman abuses the shit out of Nui's `node.extra` attribute to do basically everything so it doesn't care about the path on a node :)

**Comment by miversen33** [comment]
PR #1051 has been merged into main. I am going to close this issue out, though if my resolution did _not_ fix this, please either tag me on this so it can be reopened, or open a new issue :)
